### ISSUE TITLE
Can't browse users with large libraries

### ISSUE THREAD

**Issue by Muppetteer** [slskd/slskd#317]
When attempting to browse users with large libraries I am getting lots of errors.

First up while the browsing progress counts towards 100% there will be multiple 404 not found errors logged in the browser console like so:
```
GET http://192.168.1.3:55000/api/v0/users/username/browse/status 404 (Not Found) (anonymous) @ xhr.js:187
Uncaught (in promise) Error: Request failed with status code 404 @ createError.js:16
```

When it hits 100% the list of directories will be briefly shown before it almost instantly disappears and the following is logged in the console.
```
DOMException: Failed to execute 'setItem' on 'Storage': Setting the value of 'soulseek-example-browse-state' exceeded the quota. @ Browse.js:92
```
as a result of attempting this:
`localStorage.setItem('soulseek-example-browse-state', JSON.stringify(this.state));`

The default maximum localStorage size on Chrome and Firefox is 5mb (5120kb). For one user I was browsing, that JSON.stringify call generated over 11.8mb in a string. Perhaps you could maybe [use a compression library](https://pieroxy.net/blog/pages/lz-string/index.html) if it is necessary to store the data. That 11.8mb becomes 1.7mb using gzip on my box (though not using the linked lib). On the other hand the upper limit for a user's library size is potentially infinite so you're going to exceed this number sooner or later no matter what.

Let me know if you need the name of a user with a huge library to test on.

**Comment by Muppetteer** [comment]
Quick hack fix above. Untested and as mentioned there might be good reasons not to use it but its there if you want.

**Comment by Muppetteer** [comment]
On second thoughts this probably needs to check for compression before trying to decompress exisiting data in localStorage which won't be compressed.
[Relevant](https://stackoverflow.com/questions/65168917/javascript-lz-string-compression-check-if-string-is-already-compressed)

**Comment by jpdillingham** [comment]
Thanks for raising this; it's a problem that I've already fixed with the search and need to get around to doing the same with browse results.  I used localstorage when slskd was still a tech demo because the results weren't persisted server side.

Compression is a good idea as a stopgap; it will allow larger shares to be browsed, but there will still be problems over a certain limit.

**Comment by Muppetteer** [comment]
No problem. I've tried 0.14.1 and the fix works, but barely. I guess that I am browsing a user with a massive library and React is chugging on the sheer amount of data needed to construct the clickable list box in the DOM.

I see loads of "lost connection" messages and blank grey pages with the loading spinner while processing results. Also the same after clicking on the results to look at an individual folder. But if you have the patience to sit there and let it work it out for 5+ minutes, you can at least browse and download.

**Comment by jpdillingham** [comment]
Yeah there's definitely too much rendering going on.  I haven't spent much time optimizing React, but I think what I'll probably end up trying first is rendering the tree as a flat list and then using CSS to fake the nesting.  After that I'm sure I can find a way to only render the components that are currently in view (this gets really difficult with nested elements, I think) which should speed things up to where they need to be.

**Comment by mattgoldspink** [comment]
Just submitted a fix for this. I think this should avoid a big rewrite of the screen. There was just a huge amount of unnecessary duplicate processing. Happy to try optimize more for even larger directory scan, but what I did so far is a massive improvement for what I've been experiencing

**Comment by jpdillingham** [comment]
Re-opening since there's still work to do, but it should be better.

**Comment by jpdillingham** [comment]
Just to bump this because there have been a few issues covering the same problem; this requires a pretty heavy overhaul of the browse interface that I can't afford to do right now.

**Comment by bugmancx** [comment]
I'd like to link the userscripts I added to this issue as inspiration for this work: https://github.com/slskd/slskd/issues/1254

**Comment by gabrielpoca** [comment]
Hello, wouldn't it be sufficient to save only the search parameter to local storage instead of saving the entire state? Although this would result in more requests when navigating between pages, it should enable browsing of large folders. I can make this change if you think it's ok.

**Comment by davidarkemp** [comment]
Part of the issue of this is throwing a potentially huge JSON file at the UI. 
Testing with `cpm4`, this results is around 150MB of JSON data that needs to be parsed and reparsed. Even with a slightly more optimised `getDirectoryTree` and disabling the saveState, it takes so long, the heartbeat to the server doesn't happen and the UI thinks the backend is disconnected - you see it for a second and then it comes back to the front end.
I'd suggest that a web worker would be needed here, and potentially that could cache the state, and only return the parts of the tree to the front end that are needed for that render. It's a big job.

**Comment by solomon-b** [comment]
Hi. I'm not sure if this is the same issue but I'm getting a continuous stream of 404 errors from`GET $MYHOSTNAME/api/v0/users//browse/status` from the browse (which is stuck at 0% downloaded) page.

I tried restarting `slskd` but the requests have persisted. Is there some way to at least cancel the browse request?

**Comment by jpdillingham** [comment]
Clearing local storage as described in #1275 will likely fix this.

**Comment by solomon-b** [comment]
thank you!

**Comment by slydetector** [comment]
This is indeed quite an overhaul.  Just ask any LLM: 

Let's say I have a 100mb json file that contains a heavily nested tree structure. I want to create a responsive web based UI to browse the data in the json file. It is unreasonable for the single 100mb json file to be downloaded in one shot and rendered in a web browser like firefox. What are some proven architectural patterns to enable this sort of server to browser based interaction that minimizes data transfer and maintains a high degree of usability without taxing the browser for resources?

**Comment by slydetector** [comment]
@johnhamelink @ValerioLyndon Help me understand the reason for the 👎.

**Comment by pwall2222** [comment]
Isn't the issue the lack of virtual scroll?

**Comment by l1znix** [comment]
hey, happy to take this on if it's still up for grabs.

rough plan is basically doing for browse what you did for search :

- cache the browse result server-side (bounded, like search) and parse it into a tree once.
- new lazy, paginated endpoints so the ui only pulls the folder/page it's showing, no more dumping the whole library into the browser or localstorage.
- rewrite the browse components to lazy-load off those.
- while i'm in there: recursive directory download (walk the subtree server-side and enqueue via the existing batch path), comes almost for free and gets asked for a lot.

main tradeoff is it moves the memory cost off the browser and onto a bounded chunk of server memory, which seems in line with the direction you already took for search.

what do you think?

**Comment by jpdillingham** [comment]
Thanks @l1znix for your interest, but this is now blocked on a change to Soulseek.NET that will allow me to stream the response one directory at a time to reduce memory pressure; this work isn't worth doing unless it will fix all of the problems.

**Comment by danya02** [comment]
Hi! To be clear, the blocking issue is https://github.com/jpdillingham/Soulseek.NET/issues/889?

I suppose that making responses streamable would be helpful in general, not just for slskd but for any users of Soulseek.NET. I'm not sure if it'll solve the immediate issue: if a response to browse is indeed too huge, then the bottleneck will likely be the DOM rendering.

To deal with that, I think we should make the view emit a limited number of DOM elements. For example, say that by default you can only render 1k items (including files and folders), and if the amount to display exceeds that then you get a warning. From there you can either increase the limit (increasing the load on the browser), or filter the result set in memory with a search box.

**Comment by jpdillingham** [comment]
That's certainly the plan, but it isn't worth doing until the backend changes are completed, as I won't be able to test the solution at its limits.  

Right now there are several users on the network that are sharing enough files to cause an out of memory exception due to the number of memory allocations while handling the massive response.  There are is a much larger group of users whose shares are too large to fit in local storage in a browser, and then larger still whose shares can't be rendered in the tree and so on.

There are a number of ways to solve this in the UI, including windowing, lazy loading parts of the tree, using vanilla javascript, other ideas I'm sure folks have.  The UI isn't what is holding this up at this point.

**Comment by jpdillingham** [comment]
The screen is now much more functional:

<img width="1219" height="781" alt="Image" src="https://github.com/user-attachments/assets/2d12925b-e21a-4cb7-9090-91949fb681b7" />

I have tested it with 1 million files/100k directories and it is mostly fine.

The directory tree can be selectively expanded contracted, and I've started using `react-virtual` so that only what's visible is rendered.  Both help with the responsiveness when large shares are loaded.

I moved the storage of the actual browse response itself into IndexedDb which isn't bound by localStorage's ~5mb size limit.  I added a cap of 5000 selected files, since the selection is still stored in localStorage and the risk of overflow comes back into play.

The solution still isn't great, and is still not capable of browsing the largest shares, but it should be considerably more usable for most people.
